File: src/main/java/org/kite9/diagram/dom/processors/xpath/XPathValueReplacer.java
package org.kite9.diagram.dom.processors.xpath;

import org.kite9.diagram.dom.elements.ADLDocument;
import org.kite9.diagram.dom.processors.xpath.ValueReplacingProcessor.ValueReplacer;
import org.w3c.dom.Element;
import org.w3c.dom.xpath.XPathEvaluator;
import org.w3c.dom.xpath.XPathNSResolver;
import org.w3c.dom.xpath.XPathResult;

public class XPathValueReplacer implements ValueReplacer {
	
    final XPathEvaluator xpathEvaluator;
    final XPathNSResolver resolver;
     
    public XPathValueReplacer(ADLDocument in) {
    		this.xpathEvaluator = in;
    		this.resolver = null;
    }


	@Override
	public String getReplacementValue(String xpath, Element at) {
		return (String) xpathEvaluator.evaluate(xpath, at, resolver, XPathResult.STRING_TYPE, null);
	}

}

File: k9layout/src/main/kotlin/org/kite9/diagram/dom/processors/xpath/XPathValueReplacer.kt
package org.kite9.diagram.dom.processors.xpath

import org.kite9.diagram.dom.bridge.ElementContext
import org.w3c.dom.Element

class XPathValueReplacer(val ctx: ElementContext) : PatternValueReplacer() {

    override fun getReplacementStringValue(s: String, at: Element): String {
        val av = replaceVariables(s, at)
        val result = evaluateXPath(av, at)
        return result ?: ""
    }


    fun replaceVariables(s: String, at: Element) : String {
        val p = Regex("\\$[a-z]+")
        val out = replacePattern(p, s, at) { mr, at -> getVariableValue(mr.groupValues[0].substring(1), at) }
        return out
    }

    fun getVariableValue(v: String, at: Element) : String {
        var at : Element? = at;
        do {
            if (at is Element) {
                val de = ctx.getRegisteredDiagramElement(at)
                if (de is XPathAware) {
                    val out = de.getXPathVariable(v)
                    if (out != null) {
                        return out
                    }
                }
            }
            if (at == null) {
                return ""
            } else {
                val parent = at.parentNode
                if (parent is Element) {
                    at = parent
                } else {
                    at = null
                }
            }
        } while (at != null)

        return ""
    }

    fun evaluateXPath(s: String, at: Element) : String? {
        return ctx.evaluateXPath(s, at)
    }
}